hail/src/main/scala/is/hail/expr/ir/TableIR.scala

```diff
@@ -622,7 +623,31 @@ case class TableJoin(left: TableIR, right: TableIR, joinType: String, joinKey: I
     val rightKeyFieldIdx = rightRVDType.kFieldIdx
     val leftValueFieldIdx = leftRVDType.valueFieldIdx
     val rightValueFieldIdx = rightRVDType.valueFieldIdx
-    val newRowPType = PType.canonical(newRowType).asInstanceOf[PStruct]
+
+    val leftKeyType = leftRVDType.kType
```

Comment:
the requiredness of key fields depends on the join type too. Suppose the left key is required, right key is optional -- an inner join should return required, an outer join should return optional.

Reply:
Wasn't sure about this, because in the relational db setting 1) the notion of a primary key is that it is always present, it is the unique identifier of the row 2) one joins on keys and returns values. So values can be missing as a consequence of the kind of join, but not keys.

Reply:
missing and duplicate keys are totally supported in Hail, yep. An outer join of a table with a required key and a table with an optional key can totally have missing keys.

Reply:
I should use hail more